plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryQueryPageSource.java

```diff
@@ -146,7 +162,21 @@ public long getMemoryUsage()
     public Page getNextPage()
     {
         verify(pageBuilder.isEmpty());
-        for (FieldValueList record : tableResult.iterateAll()) {
+        if (tableResult == null) {
+            tableResult = getFutureValue(tableResultFuture);
+        }
+        else if (tableResult.hasNextPage()) {
+            long start = System.nanoTime();
+            tableResult = tableResult.getNextPage();
+            readTimeNanos.addAndGet(System.nanoTime() - start);
+        }
+        else {
+            finished = true;
+            return null;
+        }
+
+        long start = System.nanoTime();
+        for (FieldValueList record : tableResult.getValues()) {
```

Comment:
Ideally readTimeNanos shouldn't count the time taken to construct the trino page.
Maybe we write this as
```
long start = System.nanoTime();
List<FieldValueList> values = ImmutableList.copyOf(tableResult.getValues());
readTimeNanos.addAndGet(System.nanoTime() - start);
for (FieldValueList record : values) {
.......
```

Reply:
Thanks for clarifying, I was wondering about that but never asked!